doc/exts/pylint_options.py

```diff
@@ -0,0 +1,183 @@
+# Licensed under the GPL: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
+# For details: https://github.com/PyCQA/pylint/blob/main/LICENSE
+# Copyright (c) https://github.com/PyCQA/pylint/blob/main/CONTRIBUTORS.txt
+
+"""Script used to generate the options page."""
+
+from __future__ import annotations
+
+import re
+from collections import defaultdict
+from inspect import getmodule
+from pathlib import Path
+from typing import Dict, List, NamedTuple
+
+import tomlkit
+from sphinx.application import Sphinx
+
+from pylint.checkers import initialize as initialize_checkers
+from pylint.checkers.base_checker import BaseChecker
+from pylint.extensions import initialize as initialize_extensions
+from pylint.lint import PyLinter
+from pylint.typing import OptionDict
+from pylint.utils import get_rst_title
+
+
+class OptionsData(NamedTuple):
+    name: str
+    optdict: OptionDict
+    checker: BaseChecker
+    extension: bool
+
+
+OptionsDataDict = Dict[str, List[OptionsData]]
+
+PYLINT_BASE_PATH = Path(__file__).resolve().parent.parent.parent
+"""Base path to the project folder."""
+
+PYLINT_USERGUIDE_PATH = PYLINT_BASE_PATH / "doc" / "user_guide"
+"""Path to the messages documentation folder."""
+
+
+def _register_all_checkers_and_extensions(linter: PyLinter) -> None:
+    """Registers all checkers and extensions found in the default folders."""
+    initialize_checkers(linter)
+    initialize_extensions(linter)
+
+
+def _get_all_options(linter: PyLinter) -> OptionsDataDict:
+    """Get all options registered to a linter and return the data."""
+    all_options: OptionsDataDict = defaultdict(list)
+    for checker in sorted(linter.get_checkers()):
+        ch_name = checker.name
+        for option in checker.options:
+            all_options[ch_name].append(
+                OptionsData(
+                    option[0],
+                    option[1],
+                    checker,
+                    getmodule(checker).__name__.startswith("pylint.extensions."),  # type: ignore[union-attr]
+                )
+            )
+
+    return all_options
+
+
+def _create_checker_section(
+    checker: str, options: list[OptionsData], linter: PyLinter
+) -> str:
+    checker_string = get_rst_title(f"``{checker.capitalize()}`` Checker", "^")
+
+    toml_doc = tomlkit.document()
+    pylint_tool_table = tomlkit.table(is_super_table=True)
+    toml_doc.add(tomlkit.key(["tool", "pylint"]), pylint_tool_table)
+
+    checker_table = tomlkit.table()
+
+    for option in sorted(options, key=lambda x: x.name):
+        checker_string += get_rst_title(f"--{option.name}", '"')
+        checker_string += f"\nDescription: *{option.optdict.get('help')}*\n\n"
+        checker_string += f"Default: *{option.optdict.get('default')}*\n\n\n"
```

Comment:
if the default value is empty string `""`, the docs will just render as  `**`
![image](https://user-images.githubusercontent.com/55311527/163673504-56b62eaf-f135-4d41-82d4-ab9e144742d6.png)
maybe we should use `metavar` to replace this, when the default value is empty string.
I check many options that have default value `""` either accept `regex`, `msg id`, or some multiple choice
There are 13 case for this

Reply:
also for formatting maybe we can use
[Definition List](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#definition-lists)  or [Field list](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#field-lists) to get identation for separation between `Description` `Default` and their value

Reply:
# Definition list
```python
        checker_string += f"\nDescription:\n  *{option.optdict.get('help')}*\n\n"
        if option.optdict.get("default") == "":
            checker_string += f"Default:\n  ``{option.optdict.get('metavar')}``\n\n\n"
        else:
            checker_string += f"Default:\n  ``{option.optdict.get('default')}``\n\n\n"
```
![image](https://user-images.githubusercontent.com/55311527/163675003-2b8d3c7b-b0fb-455c-b055-451bb505f2ef.png)

Reply:
I have changed it now print `""` if that is the default value. Let me know what you think!

Reply:
looks good 👍, also another question I see a lot of similarities with `Technical Reference/Pylint features` do we need to do something about it?

Reply:
We should probably revisit the Technical Reference at some point as there is indeed a bit of duplication there. However, I don't think normal users are likely to go there to look for configuration help so I think it is good to create this new section.